lambdas/service/app.py

```diff
@@ -187,8 +187,133 @@ def version():
+        try:
+            relation, value = one(filter_.items())
+        except Exception:
+            raise BadRequestError(f'The `filters` parameter entry for `{facet}` must be a single-item dictionary.')
+        else:
+            valid_relations = ('is', 'contains', 'within', 'intersects')
+            if relation in valid_relations:
+                if not isinstance(value, list):
+                    raise BadRequestError(
+                        msg=f'Invalid type for "{relation}" in `{facet}`, should be wrapped in a list')
+            else:
+                raise BadRequestError(f'The `filters` parameter entry for `{facet}` must be one of {valid_relations}"')
+
+
+def validate_facet(value):
+    """
+    >>> validate_facet('fileName')
+
+    >>> validate_facet('fooBar')
+    Traceback (most recent call last):
+    ...
+    chalice.app.BadRequestError: BadRequestError: Invalid parameter `fooBar`
+    """
+    translation = Plugin.load().service_config().translation
+    if value not in translation:
+        raise BadRequestError(msg=f'Invalid parameter `{value}`')
+
+
+def validate_params(query_params: Mapping[str, str],
+                    allow_extra_params=False,
+                    **validators: Callable[[Any], Any]) -> None:
+    """
+    Validates request query parameters for web-service API.
+
+    :param query_params: Parameters to be validated.
+
+    :param allow_extra_params: When False, only parameters specified via '**validators' are accepted, and validation
+                               fails if additional parameters are present. When True, additional parameters are allowed
+                               but their value is not validated.
+
+    :param validators: A dictionary mapping the name of a parameter to a function that will be used to validate the
+                       parameter. The callable called with a single argument, the parameter value to be validated, and
+                       is expected to raise an exception if the value is invalid. It may return a value but
+
+    >>> validate_params({'order': 'asc'}, order=str)
+
+    >>> validate_params({'size': 'foo'}, size=int)
+    Traceback (most recent call last):
+        ...
+    chalice.app.BadRequestError: BadRequestError: Invalid input type for `size`
+
+    >>> validate_params({'order': 'asc', 'token': 'bar'}, order=str)
+    Traceback (most recent call last):
+        ...
+    chalice.app.BadRequestError: BadRequestError: Invalid query parameter `token`
+
+    >>> validate_params({'order': 'asc', 'token': 'bar'}, allow_extra_params=True, order=str)
+    """
+    for param_name, param_value in query_params.items():
+        try:
+            validator = validators[param_name]
+        except KeyError:
+            if not allow_extra_params:
+                raise BadRequestError(msg=f'Invalid query parameter `{param_name}`')
+        else:
+            try:
+                validator(param_value)
+            except (TypeError, ValueError):
```

Comment:
This is inconsistent with the docstring. I'm ok with restricting the set of exceptions that validators can raise for invalid input but that should be documented and it should be documented what happens if the validator raises an exception that is of neither of those types.